[SPARK-34382][SQL] Support LATERAL subqueries

allisonwang-db opened the pull request:
### What changes were proposed in this pull request?
This PR adds support for lateral subqueries. A lateral subquery is a subquery preceded by the `LATERAL` keyword in the FROM clause of a query that can reference columns in the preceding FROM items. For example:
```sql
SELECT * FROM t1, LATERAL (SELECT * FROM t2 WHERE t1.a = t2.c)
```
A new subquery expression`LateralSubquery` is used to represent a lateral subquery. It is similar to `ScalarSubquery` but can return multiple rows and columns. A new logical unary node `LateralJoin` is used to represent a lateral join.

Here is the analyzed plan for the above query:
```scala
Project [a, b, c, d]
+- LateralJoin lateral-subquery [a], Inner
   :  +- Project [c, d]
   :     +- Filter (outer(a) = c)
   :        +- Relation [c, d]
   +- Relation [a, b]
```

Similar to a correlated subquery, a lateral subquery can be viewed as a dependent (nested loop) join where the evaluation of the right subtree depends on the current value of the left subtree.  The same technique to decorrelate a subquery is used to decorrelate a lateral join:
```scala
Project [a, b, c, d]
+- LateralJoin lateral-subquery [a && a = c], Inner  // pull up correlated predicates as join conditions
   :  +- Project [c, d]
   :     +- Relation [c, d]
   +- Relation [a, b]
```
Then the lateral join can be rewritten into a normal join:
```scala
Join Inner (a = c)
:- Relation [a, b]
+- Relation [c, d]
```

#### Follow-ups:
1. Similar to rewriting correlated scalar subqueries, rewriting lateral joins is also subject to the COUNT bug (See SPARK-15370 for more details). This is **not** handled in the current PR as it requires a sizeable amount of refactoring. It will be addressed in a subsequent PR (SPARK-35551).
2. Currently Spark does use outer query references to resolve star expressions in subqueries. This is not lateral subquery specific and can be handled in a separate PR (SPARK-35618)

### Why are the changes needed?
To support an ANSI SQL feature.

### Does this PR introduce _any_ user-facing change?
Yes. It allows users to use lateral subqueries in the FROM clause of a query.

### How was this patch tested?
- Parser test: `PlanParserSuite.scala`
- Analyzer test: `ResolveSubquerySuite.scala`
- Optimizer test: `PullupCorrelatedPredicatesSuite.scala`
- SQL test: `join-lateral.sql`, `postgreSQL/join.sql`

allisonwang-db committed:
lateral join

SparkQA commented:
**[Test build #137824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137824/testReport)** for PR 32303 at commit [`8f6ea0a`](https://github.com/apache/spark/commit/8f6ea0a385beda190f93b7edff587d4d4e2aef18).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42354/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42354/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42354/

SparkQA commented:
**[Test build #137824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137824/testReport)** for PR 32303 at commit [`8f6ea0a`](https://github.com/apache/spark/commit/8f6ea0a385beda190f93b7edff587d4d4e2aef18).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ExpandedStar(star: UnresolvedStar) extends Star with Unevaluable `
  * `case class LateralJoin(tpe: JoinType) extends JoinType `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137824/

allisonwang-db commented:
cc @cloud-fan

cloud-fan commented:
cc @maropu @viirya

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1638,6 +1660,27 @@ class Analyzer(override val catalogManager: CatalogManager)
+   * Optionally resolve the name parts using the outer query plan and wrap resolved attributes
+   * with [[OuterReference]]s.
+   */
+  private def resolveOuterReference(
```
Just a question; instead of using the outer reference, we cannot simply pull up a filter condition into a join condition?
For example, in a query below, can an analyzer converts`'Join LateralJoin(Inner)` into `'Join LateralJoin(Inner), 't1.c1 = c1#171`, then resolves `'t1.c1` by `ResolveReferences`?
```
 +- 'Join LateralJoin(Inner)
    :- SubqueryAlias spark_catalog.default.t1
    :  +- View (`default`.`t1`, [c1#167,c2#168])
    :     +- Project [cast(col1#169 as int) AS c1#167, cast(col2#170 as int) AS c2#168]
    :        +- LocalRelation [col1#169, col2#170]
    +- 'SubqueryAlias __auto_generated_subquery_name
       +- 'Project ['c2]
          +- 'Filter ('t1.c1 = c1#171)
             +- SubqueryAlias spark_catalog.default.t2
                +- View (`default`.`t2`, [c1#171,c2#172])
                   +- Project [cast(col1#173 as int) AS c1#171, cast(col2#174 as int) AS c2#172]
                      +- LocalRelation [col1#173, col2#174]
```

allisonwang-db replied:
Pulling up correlated conditions can be tricky. In this case, it works because it's a simple equality condition, but for other cases such as having correlated non-equality conditions below an aggregate, decorrelation becomes more complex, and directly pulling up the filter conditions as join conditions may yield incorrect results (SPARK-17348 and SPARK-35080).

maropu commented:
It would be nice to add a page for the literal join in follow-up PRs.

cloud-fan commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -588,7 +588,7 @@ hintStatement
 
 fromClause
-    : FROM relation (',' relation)* lateralView* pivotClause?
+    : FROM relation (',' lateralClause)* (',' relation)* lateralView* pivotClause?
```
Interesting. But what outer references we can use? It's the leftmost query...

maropu replied:
Yea, it seems PosgreSQL accepts simple cases without outer refs.

viirya replied:
I think it was mentioned in the [manual](https://www.postgresql.org/docs/current/queries-table-expressions.html#QUERIES-LATERAL):

> A LATERAL item can appear at top level in the FROM list, or within a JOIN tree. In the latter case it can also refer to any items that are on the left-hand side of a JOIN that it is on the right-hand side of.

allisonwang-db commented on sql/core/src/test/resources/sql-tests/inputs/join-lateral.sql:
```diff
@@ -0,0 +1,84 @@
+CREATE VIEW t2(c1, c2) AS VALUES (0, 2), (0, 3);
+
+-- lateral join with single column select
+SELECT * FROM t1, LATERAL (SELECT c1);
```
> Is this compiled into a join query? We cannot compile it into a simple scan query like PostgreSQL?
> 
> ```
> postgres=# explain SELECT * FROM t1, LATERAL (SELECT c1) t2;
>                           QUERY PLAN                          
> --------------------------------------------------------------
>  Values Scan on "*VALUES*"  (cost=0.00..0.03 rows=2 width=12)
> (1 row)
> ```

Yes it should be rewritten as a simple scan query`SELECT c1 FROM t1` instead of a join query. I am thinking about adding an optimizer rule for OneRowRelation. This can be applied to correlated subqueries as well.

allisonwang-db replied:
Postgres requires all subqueries in the FROM clause to have aliases, not just lateral, but aliases are not required for subqueries in Spark.
```
select * from (select * from t1);

subquery in FROM must have an alias
```

allisonwang-db commented:
> It would be nice to add a doc page for the literal join in follow-up PRs.

Sounds good. Created SPARK-35208 as a follow up task.

allisonwang-db committed:
address comments and activate postgreSQL tests

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2234,6 +2270,74 @@ class Analyzer(override val catalogManager: CatalogManager)
+              withPosition(u) {
+                p.resolveChildren(nameParts, resolver)
+                  .orElse(resolveLiteralFunction(nameParts))
+                  .orElse(resolveOuterReference(nameParts, left))
```
It looks you forget to remove off the outer reference from the an optimized plan and the ref is unevaluable. 
```
scala> sql("SELECT * FROM t1, LATERAL (SELECT c2 FROM t2 WHERE t1.c1 = t2.c1) tx(a);").explain(true)
...
== Optimized Logical Plan ==
Join Inner
:- LocalRelation [c1#323, c2#324]
+- Project [c2#328 AS a#331]
   +- Filter (outer(c1#323) = c1#327)   <---- !!!HERE!!!
      +- LocalRelation [c1#327, c2#328]
```

allisonwang-db replied:
Hmm I am getting this optimized plan from the query:
```
scala> sql("SELECT * FROM t1, LATERAL (SELECT c2 FROM t2 WHERE t1.c1 = t2.c1) tx(a);").explain(true)
...
== Optimized Logical Plan ==
Project [c1#0, c2#1, a#8]
+- Join Inner, (c1#0 = c1#4)
   :- LocalRelation [c1#0, c2#1]
   +- LocalRelation [a#8, c1#4]
``` 
The outer references inside the Filter should have been pulled up as join conditions.

allisonwang-db replied:
I rebased the master and this is due to not adding a proper tree pattern. Should be fixed now :)

maropu commented:
```
scala> sql("SELECT * FROM (VALUES (1), (2)) tx(a), LATERAL (SELECT a FROM tx) ty(a)").show()
org.apache.spark.sql.AnalysisException: Table or view not found: tx; line 1 pos 62;
'Project [*]
+- 'Join LateralJoin(Inner)
   :- SubqueryAlias tx
   :  +- Project [col1#345 AS a#346]
   :     +- LocalRelation [col1#345]
   +- 'SubqueryAlias ty
      +- 'UnresolvedSubqueryColumnAliases [a]
         +- 'Project ['a]
            +- 'UnresolvedRelation [tx], [], false
```
How bout the case above? The table resolution couldn't work well.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2234,6 +2270,74 @@ class Analyzer(override val catalogManager: CatalogManager)
+      }
+
+      buildExpandedProjectList(
+        buildExpandedProjectList(expressions, inner, expandInner),
```
If inner `buildExpandedProjectList` was fully resolved, do we still need run outer `buildExpandedProjectList`?

allisonwang-db replied:
Good point. If there is no star in the expanded list using the inner plan, then we don't need to expand it again using the outer plan.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -857,12 +865,21 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
+
     // Make sure a plan's expressions do not contain :
     // 1. Aggregate expressions that have mixture of outer and local references.
     // 2. Expressions containing outer references on plan nodes other than Filter.
```
Also update this comment?

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:
```diff
@@ -156,6 +156,16 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
+      case oldVersion @ Project(projectList, _)
+          if findOuterReferences(projectList).intersect(conflictingAttributes).nonEmpty =>
+        // Add alias to conflicting outer references.
+        val aliasedProjectList = projectList.map {
+          case o @ OuterReference(a) if conflictingAttributes.contains(a) => Alias(o, a.name)()
+          case other => other
+        }
```
Could outer reference be inside an expression tree?

allisonwang-db replied:
Outer references can be inside an expression tree, which can be handled by the deduplicating aliases case, for example `outer(a) AS a`. Here the dedup logic is trying to handle this case:
```
Join LatealJoin(Inner)
:- Project [c1#0, c2#0]
:-  ...
+- Project [outer(c1#0)]
   +- ...
```
Since an outer reference is also a named expression, it can appear in the project list without an alias. Then `c1#0` and  `outer(c1#0)` will conflict with each other. (`outer(c1#0).toAttribute => c1#0`)

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -392,16 +392,27 @@ case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevalu
 }
 
+/**
+ * Represent an [[UnresolvedStar]] that has been to expand once.
```
`has been to expand once` sounds uncertain definition to me. By its definition, cannot it be expanded more than once?

allisonwang-db replied:
This `ExpandedStar` is indeed confusing. It was used to detect invalid usage of star expressions. I will update the logic here.

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -392,16 +392,27 @@ case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevalu
+ * Represent an [[UnresolvedStar]] that has been to expand once.
+ * This can be used to resolve a star expression using multiple plans.
+ */
+case class ExpandedStar(star: UnresolvedStar) extends Star with Unevaluable {
```
This is used to differenate between an unresolved star and a star that has been tried to resolved once. It's intended to keep the error message consistent when the star usage is invalid in an expression. But indeed this is confusing and I will think of other ways to check the invalid usage.

SparkQA commented:
**[Test build #137967 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137967/testReport)** for PR 32303 at commit [`07d764a`](https://github.com/apache/spark/commit/07d764a3593888bc9c7b6ad3ed03733c56212925).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42487/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42487/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42487/

allisonwang-db commented:
> ```
> scala> sql("SELECT * FROM (VALUES (1), (2)) tx(a), LATERAL (SELECT a FROM tx) ty(a)").show()
> org.apache.spark.sql.AnalysisException: Table or view not found: tx; line 1 pos 62;
> 'Project [*]
> +- 'Join LateralJoin(Inner)
>    :- SubqueryAlias tx
>    :  +- Project [col1#345 AS a#346]
>    :     +- LocalRelation [col1#345]
>    +- 'SubqueryAlias ty
>       +- 'UnresolvedSubqueryColumnAliases [a]
>          +- 'Project ['a]
>             +- 'UnresolvedRelation [tx], [], false
> ```
> 
> How bout the case above? The table resolution couldn't work well.

In this case `tx` in `LATERAL (SELECT a FROM tx)` will be resolved as a relation instead of a lateral reference. Postgres also raises an error in this case:
```
SELECT * FROM (VALUES (1), (2)) tx(a), LATERAL (SELECT a FROM tx) ty(a)
relation "tx" does not exist
```

Both of the following queries will work:
```sql
SELECT * FROM (VALUES (1), (2)) tx(a), LATERAL (SELECT a) ty(a)
SELECT * FROM (VALUES (1), (2)) tx(a), LATERAL (SELECT tx.a) ty(a)
```

SparkQA commented:
**[Test build #137967 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137967/testReport)** for PR 32303 at commit [`07d764a`](https://github.com/apache/spark/commit/07d764a3593888bc9c7b6ad3ed03733c56212925).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137967/

allisonwang-db committed:
update tree pattern

SparkQA commented:
**[Test build #137973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137973/testReport)** for PR 32303 at commit [`af0c3b0`](https://github.com/apache/spark/commit/af0c3b0bb4647ce50700151f6c04089c34f915d9).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42493/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42493/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42493/

SparkQA commented:
**[Test build #137973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137973/testReport)** for PR 32303 at commit [`af0c3b0`](https://github.com/apache/spark/commit/af0c3b0bb4647ce50700151f6c04089c34f915d9).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137973/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2234,6 +2260,76 @@ class Analyzer(override val catalogManager: CatalogManager)
+    }
+
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case j @ Join(left, right, LateralJoin(_), _, _) if left.resolved && !right.resolved =>
```
shall we also make sure `right` doesn't contain unresolved lateral join?

allisonwang-db replied:
Good point. Right now the decorrelation framework won't be able to support deep correlation for example
```sql
select * from t1, lateral (select * from t2, lateral (select t1.c1 + t2.c1));
```
But this can be resolved successfully in the Analyzer.

allisonwang-db replied:
This is trickier than expected. We can't simply check that `right` doesn't contain unresolved lateral join because the left subtree of a nested lateral join can contain outer references:
```sql
SELECT * FROM t1, LATERAL (SELECT * FROM (SELECT t1.c1 + 1 AS c1), LATERAL (SELECT c1))
```
Plan
```scala
Join LateralJoin (1)
:- Relation T1
+- Join LateralJoin (2)
   :- Project [unresolvedalias(t1.c1 + 1, c1)]
   :  +- OneRowRelation
   +- Project ['c1]
      +- OneRowRelation
```
Here the outermost lateral join's right subtree will get resolved first, which will resolve `'c1` as `t1.c1`. But in this case `c1` should be resolved as `t1.c1 + a AS c1` from the inner lateral join's left subtree instead of `t1.c1` (double checked with Postgres). I will update the resolution logic here.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -1037,6 +1057,10 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
 
+          // Lateral join is not supported inside a correlated subquery.
+          case LateralJoin(_) =>
+            failOnOuterReferenceInSubTree(j)
```
what's the error message? Do we clearly mention that lateral join inside correlated subquery is not allowed?

allisonwang-db replied:
The error message here is "Accessing outer query column is not allowed in LateralJoin...". I will make the error message more explicit.

allisonwang-db committed:
nested lateral join

SparkQA commented:
**[Test build #138059 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138059/testReport)** for PR 32303 at commit [`2e87e18`](https://github.com/apache/spark/commit/2e87e18351a1eea358fcb98090deb3a11e839059).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42578/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42578/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42578/

SparkQA commented:
**[Test build #138059 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138059/testReport)** for PR 32303 at commit [`2e87e18`](https://github.com/apache/spark/commit/2e87e18351a1eea358fcb98090deb3a11e839059).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138059/

allisonwang-db commented:
@maropu @viirya @cloud-fan Thanks for the review! I've addressed the comments and activated some Postgres SQL tests for lateral subqueries. The results also match Postgres' results.

allisonwang-db committed:
update analyzer

SparkQA commented:
**[Test build #138341 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138341/testReport)** for PR 32303 at commit [`f726c83`](https://github.com/apache/spark/commit/f726c8375e70f3fb984276e713fca747fe765f45).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42863/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42863/

SparkQA commented:
**[Test build #138341 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138341/testReport)** for PR 32303 at commit [`f726c83`](https://github.com/apache/spark/commit/f726c8375e70f3fb984276e713fca747fe765f45).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138341/

SparkQA commented:
**[Test build #138348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138348/testReport)** for PR 32303 at commit [`62d8a97`](https://github.com/apache/spark/commit/62d8a9756239ba92129ce6a20d037aa00b0707f1).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42870/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42870/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138348/

SparkQA commented:
**[Test build #138764 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138764/testReport)** for PR 32303 at commit [`fe2f941`](https://github.com/apache/spark/commit/fe2f9415fe020acd214557327cf17bd1174c1bf2).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43286/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43286/

SparkQA commented:
**[Test build #138764 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138764/testReport)** for PR 32303 at commit [`fe2f941`](https://github.com/apache/spark/commit/fe2f9415fe020acd214557327cf17bd1174c1bf2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138764/

SparkQA commented:
**[Test build #138817 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138817/testReport)** for PR 32303 at commit [`8e6cab3`](https://github.com/apache/spark/commit/8e6cab33e701a55d698a7239947ab85423010b7e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43339/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43339/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43339/

SparkQA commented:
**[Test build #138817 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138817/testReport)** for PR 32303 at commit [`8e6cab3`](https://github.com/apache/spark/commit/8e6cab33e701a55d698a7239947ab85423010b7e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138817/

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2395,6 +2421,70 @@ class Analyzer(override val catalogManager: CatalogManager)
+  /**
+   * This rule resolves lateral joins.
+   */
+  object ResolveLateralJoin extends Rule[LogicalPlan] {
```
`ResolveReferencesInLateralJoin`?

maropu replied:
Q: Btw, is it difficult to merge this resolution logic with the `ResolveReferences` one?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -276,6 +276,7 @@ class Analyzer(override val catalogManager: CatalogManager)
       ResolveAliases ::
       ResolveSubquery ::
       ResolveSubqueryColumnAliases ::
+      ResolveLateralJoin ::
```
Since `ResolveLateralJoin` has a similar functionality with `ResolveReferences`, how about putting this rule just after `ResolveReferences`?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -174,13 +174,21 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
             }
+          // Check if there are expressions with children containing star expressions.
+          case e if e.children.exists(_.isInstanceOf[Star]) =>
+            e.failAnalysis(s"Invalid usage of '*' in expression '${e.prettyName}'")
```
If we add this rule here, I think we need to update the comment above.
https://github.com/apache/spark/pull/32303/files#diff-583171e935b2dc349378063a5841c5b98b30a2d57ac3743a9eccfe7bffcb8f2aL165-L168

allisonwang-db replied:
Good point. The error message is slightly different from `invalidStarUsageError`(with an extra word 'expression') but it should be fine to unify them.

maropu commented on sql/core/src/test/resources/sql-tests/inputs/join-lateral.sql:
```diff
@@ -0,0 +1,97 @@
+SELECT * FROM t1, LATERAL (SELECT t1.c1 FROM t2);
+SELECT * FROM t1, LATERAL (SELECT t1.c1 + t2.c1 FROM t2);
+
+-- lateral join with star expansion
```
Could you add tests for `quotedRegexColumnNames`, too?

allisonwang-db replied:
All star expressions will be handled in a separate PR: SPARK-35618

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -174,13 +174,21 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
             val from = operator.inputSet.toSeq.map(_.qualifiedName).mkString(", ")
             a.failAnalysis(s"cannot resolve '${a.sql}' given input columns: [$from]")
 
+          case s @ UnresolvedStar(Some(target)) =>
```
We still need this error check even if we check it in L178?

allisonwang-db replied:
Yup L178 checks invalid star usage inside expressions first, for example, `sum(t.*)` and here the unresolved star is not inside any expressions.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -254,6 +262,27 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
+                    val column = outer.outputSet.toSeq.map(_.qualifiedName).mkString(", ")
+                    o.failAnalysis(
+                      s"Found an outer column reference '${e.sql}' in a lateral subquery " +
+                      s"that is not present in the preceding FROM items of its own query " +
```
nit: remove `s`

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:
```diff
@@ -182,6 +182,16 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
         Seq((oldVersion, oldVersion.copy(projectList = newAliases(projectList))))
 
+      // Handle projects that create conflicting outer references.
+      case oldVersion @ Project(projectList, _)
```
cc: @Ngone51

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -402,10 +402,8 @@ case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevalu
     } else {
-      val from = input.inputSet.map(_.name).mkString(", ")
-      val targetString = target.get.mkString(".")
-      throw QueryCompilationErrors.cannotResolveStarExpandGivenInputColumnsError(
```
It seems no one uses this method now?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala:
```diff
@@ -244,7 +244,8 @@ object DecorrelateInnerQuery extends PredicateHelper {
             case _ => Join(child, domain, Inner, None, JoinHint.NONE)
           }
         } else {
-          throw QueryExecutionErrors.cannotRewriteDomainJoinWithConditionsError(conditions, d)
```
ditto: It seems no one uses this method now?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:
```diff
@@ -168,6 +168,21 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
+    case j @ Join(left, right, LateralJoin(joinType), condition, _) =>
+      val conditions = condition.map(splitConjunctivePredicates).getOrElse(Nil)
+      val newRight = DecorrelateInnerQuery.rewriteDomainJoins(left, right, conditions)
+      // TODO: handle the COUNT bug
```
Can you add the jira number (SPARK-15370) here?

allisonwang-db replied:
Created a new ticket for handling the COUNT bug in lateral subqueries: https://issues.apache.org/jira/browse/SPARK-35551

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala:
```diff
@@ -107,6 +107,11 @@ case class UsingJoin(tpe: JoinType, usingColumns: Seq[String]) extends JoinType
 }
 
+case class LateralJoin(tpe: JoinType) extends JoinType {
+  require(Seq(Inner, LeftOuter, Cross).contains(tpe), "Unsupported lateral join type " + tpe)
```
> Currently INNER, CROSS and LEFT OUTER join types are supported with lateral join

If we can support the other join types in future, how about checking this requirement in the `AstBuilder` side?

allisonwang-db replied:
@maropu Postgres supports INNER, CROSS, and LEFT lateral join, and it doesn't make sense to support RIGHT OUTER and FULL OUTER lateral join. How about let's add the other two types of supported left joins: left semi and left anti here. Then lateral join types shouldn't be changed in the future.

maropu commented on sql/core/src/test/resources/sql-tests/inputs/postgreSQL/groupingsets.sql:
```diff
@@ -499,14 +499,14 @@ SELECT a, b, count(*), max(a), max(b) FROM gstest3 GROUP BY GROUPING SETS(a, b,(
 -- Rescan logic changes when there are no empty grouping sets, so test
 -- that too:
--- [SPARK-27877] ANSI SQL: LATERAL derived table(T491)
+-- Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses:
```
We can support this case in future? If we can, can you file jira for it and update the jira number above (`[SPARK-27877] ANSI SQL: LATERAL derived table(T491)`)?

maropu reviewed (commented):
It seems the relevant errors happened in the tests.

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1607,33 +1622,30 @@ class Analyzer(override val catalogManager: CatalogManager)
           })
-        // count(*) has been replaced by count(1)
-        case o if containsStar(o.children) =>
-          throw QueryCompilationErrors.invalidStarUsageError(s"expression '${o.prettyName}'")
```
Yes it was moved to CheckAnalysis because the star expression might be resolved using outer plan references.

allisonwang-db committed:
unify subquery and lateral join

SparkQA commented:
**[Test build #139250 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139250/testReport)** for PR 32303 at commit [`72facb0`](https://github.com/apache/spark/commit/72facb0de716ed18b087017539e9dfb4dcd47d7a).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43774/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43774/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43774/

SparkQA commented:
**[Test build #139250 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139250/testReport)** for PR 32303 at commit [`72facb0`](https://github.com/apache/spark/commit/72facb0de716ed18b087017539e9dfb4dcd47d7a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class LateralSubquery(`
  * `case class LateralJoin(`

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139250/

allisonwang-db committed:
some clean up

SparkQA commented:
**[Test build #139269 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139269/testReport)** for PR 32303 at commit [`b83c571`](https://github.com/apache/spark/commit/b83c57108d6cfb839128e8e554d8f2f51906cf83).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43792/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43792/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43792/

SparkQA commented:
**[Test build #139269 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139269/testReport)** for PR 32303 at commit [`b83c571`](https://github.com/apache/spark/commit/b83c57108d6cfb839128e8e554d8f2f51906cf83).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139269/

allisonwang-db commented:
@maropu @cloud-fan @viirya 
I updated the logical plan structure to handle a few cases I found during the test that do not work with the original plan representation. Originally, we directly use the logical Join node with a new join type `LateralJoin`. This structure can't handle the case when deduping relations, for instance:
```scala
'Join
:- LocalRelation [a, b]
+- Join LateralJoin(Inner)
   :- LocalRelation [a, b]  <--- after DeduplicateRelation: LocalRelation ['a, 'b]
   +- Project [outer(a) + outer(b)]  <--- but outer references won't be updated
      +- OneRowRelation
```
`DeduplicateRelation` rule will return a new instance of the `LocalRelation[a, b]`, but the current `transformUpWithNewOutput` won't be able to update the existing outer references in the right subtree. Also, the resolution logic has a lot of overlaps with `ResolveSubquery`. The new logic unifies the analysis logic for lateral subqueries and existing subqueries. It adds a new unary node `LateralJoin` and a new subquery expression `LateralSubquery` to represent a lateral subquery. I've updated the PR description. 

Also, I removed the logic to resolve star expressions in subqueries. It is not lateral subquery specific and I've created a new JIRA ticket: SPARK-35618.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2364,11 +2364,20 @@ class Analyzer(override val catalogManager: CatalogManager)
+        val newRight = resolveSubQuery(right, j.children)(LateralSubquery(_, _, right.exprId))
+          .asInstanceOf[LateralSubquery]
+        val newCond = if (newRight.resolved && condition.exists(!_.resolved)) {
+          condition.map(resolveExpressionByPlanOutput(_, newRight.plan))
```
It's ok to have special column resolution logic. If you checkout the rule `ResolveReferences`, there are many special cases.

Let's put the special case for `LateralJoin` in `ResolveReferences` as well, to centralize all the column resolution logic. Let's also follow the general column resolution rule: fail if the name is ambiguous. For example, if join condition refers to a column in both left and right sides (un-qualified column name), we should fail.

cloud-fan replied:
How about we overwrite `resolveChildren` in `LateralJoin`, so that we consider input columns from both left and right tables? Then we don't need the special logic here, and we also follow normal join behavior and fail analysis if the name is ambiguous (`t1 JOIN LATERAL t2` ON a = 1 while both t1 and t2 have column a).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala:
```diff
@@ -244,7 +244,17 @@ object DecorrelateInnerQuery extends PredicateHelper {
+          // +- Project [(c1#270 + 1) AS c1#262, c1#270]
+          //    +- DomainJoin [c1#270]
+          //       +- OneRowRelation
+          d
```
what will we do later to resolve it?

allisonwang-db replied:
It will be rewritten during the optimization for the outer lateral joins. You can think of each DomainJoin has a "scope" defined, i.e which outer plans it belongs to, and here we only rewrite a domain join if it belongs to the current outer plan.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -285,6 +285,39 @@ object ScalarSubquery {
+ */
+case class LateralSubquery(
+    plan: LogicalPlan,
+    outerAttrs: Seq[Expression] = Seq.empty,
```
can this be `Seq[Attribute]`?

cloud-fan replied:
I see, this is the outer references, the name follows `ScalarSubquery`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -315,19 +314,28 @@ object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper
     case q: UnaryNode =>
-      rewriteSubQueries(q, q.children)
+      val newPlan = rewriteSubQueries(q, q.children)
+      // Preserve the original output of the node.
```
when can this happen?

allisonwang-db replied:
After decorrelation, the right child can have domain attributes preserved which can change its output, for example
 
// lateral subquery (before)
Project [a, b]
+- Filter [outer(c) = c]
      +- LocalRelation [a, b, c]

// lateral subquery (after decorrelate)
Project [a, b, c]
+- LocalRelation [a, b, c]

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala:
```diff
@@ -45,6 +45,7 @@ object PlanHelper {
         case e: AggregateExpression
           if !(plan.isInstanceOf[Aggregate] ||
                plan.isInstanceOf[Window] ||
+               plan.isInstanceOf[LateralJoin] ||
```
`LateralJoin` can host aggregate expressions?

allisonwang-db replied:
This is because the children (outerAttrs) field can contain aggregate expressions `LateralJoin(left, LateralSubquery(plan, [sum(a)]), ..)` when collecting outer references in the subquery:

https://github.com/apache/spark/blob/dfd8a8dc676c388c0c1bb7e4cb8d55eab10504ab/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala#L201-L211

maropu replied:
How about more strictly checking that the right plan exprs of the lateral join are aggr ones?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -871,7 +871,13 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
-      val join = right.optionalMap(left)(Join(_, _, Inner, None, JoinHint.NONE))
+      val join = right.optionalMap(left) { (left, right) =>
+        if (relation.LATERAL != null) {
+          LateralJoin(left, LateralSubquery(right), Inner, None)
```
We do not handle join hints for the lateral case?

allisonwang-db replied:
Good point. I will add it.

allisonwang-db replied:
Actually, does it make sense to have join hints for lateral joins? A lateral join is essentially a nested loop join. Ideally, the evaluation logic should be for each row in the left, plug the outer query attribute values into the outer references and evaluate the subquery. So it should only be planned as a (correlated) nested loop join. But since Spark doesn't support such execution, it first decorrelates the subquery and then rewrites the lateral join as a normal join. This seems to be an implementation detail and it doesn't make sense to add join hints here.

allisonwang-db replied:
Sounds good.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -285,6 +285,39 @@ object ScalarSubquery {
+  override def dataType: DataType = plan.output.toStructType
+  override def nullable: Boolean = true
+  override def withNewPlan(plan: LogicalPlan): LateralSubquery = copy(plan = plan)
+  override def toString: String = s"lateral-subquery#${exprId.id} $conditionString"
```
Could you add simple tests for checking if an explain can print this new node correctly?

maropu reviewed (commented):
It looks almost good and I left minor comments.

allisonwang-db committed:
address comments

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -315,19 +314,28 @@ object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper
-      rewriteSubQueries(q, q.children)
+      val newPlan = rewriteSubQueries(q, q.children)
+      // Preserve the original output of the node.
+      if (newPlan.output != q.output) {
```
Nice!

SparkQA commented:
**[Test build #139452 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139452/testReport)** for PR 32303 at commit [`d646720`](https://github.com/apache/spark/commit/d646720edf977ea50ac8f273eea75b045915038e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43975/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43975/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43975/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2357,18 +2356,22 @@ class Analyzer(override val catalogManager: CatalogManager)
       // its child for resolution.
       case f @ Filter(_, a: Aggregate) if f.childrenResolved =>
         resolveSubQueries(f, Seq(a, a.child))
+      case j @ LateralJoin(left, right, _, _) if left.resolved && !right.resolved =>
```
nit: do we need `!right.resolved`? Other cases don't have it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala:
```diff
@@ -217,36 +217,36 @@ object DecorrelateInnerQuery extends PredicateHelper {
+          // Join Inner
+          // :- Inner Query
+          // +- Domain
+          case _ => Join(child, domain, Inner, None, JoinHint.NONE)
```
shall we add an assert that: `child` has no more `DomainJoin`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -1419,3 +1419,58 @@ case class DomainJoin(domainAttrs: Seq[Attribute], child: LogicalPlan) extends U
+      resolver: Resolver): Option[NamedExpression] = {
+    val children = left :: right.plan :: Nil
+    val childAttributes = AttributeSeq(children.flatMap(_.output))
+    val childMetadataAttributes = AttributeSeq(children.flatMap(_.metadataOutput))
```
to speed this up, can we add
```
private[this] lazy val childAttributes = AttributeSeq(...)

private[this] lazy val childMetadataAttributes = AttributeSeq(...)
```

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveSubquerySuite.scala:
```diff
@@ -53,4 +58,117 @@ class ResolveSubquerySuite extends AnalysisTest {
+    // SELECT * FROM t1, LATERAL (SELECT a, c)
+    assertAnalysisError(
+      LateralJoin(t1, LateralSubquery(t0.select('a, 'c)), Inner, None),
+      Seq("cannot resolve 'c' given input columns: []")
```
note: we can improve the error message in the future, to include columns from the outer plan as well.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveSubquerySuite.scala:
```diff
@@ -53,4 +58,117 @@ class ResolveSubquerySuite extends AnalysisTest {
+      LateralJoin(t4, LateralSubquery(t0.select($"x.a", $"y.a")), Inner, None),
+      LateralJoin(t4,
+        LateralSubquery(
+          Project(Seq(OuterReference(xa.as(a.name)), OuterReference(ya.as(a.name))), t0),
```
should the outer reference wrap the attribute instead of `GetStructField`?

cloud-fan reviewed (approved):
LGTM except for some minor comments

SparkQA commented:
**[Test build #139452 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139452/testReport)** for PR 32303 at commit [`d646720`](https://github.com/apache/spark/commit/d646720edf977ea50ac8f273eea75b045915038e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139452/

allisonwang-db committed:
address comments

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44035/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44035/

cloud-fan committed:
Update Analyzer.scala

SparkQA commented:
**[Test build #139528 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139528/testReport)** for PR 32303 at commit [`8f19ab7`](https://github.com/apache/spark/commit/8f19ab732fa0117a72626c601e964140c9b4bd42).

allisonwang-db committed:
update test

SparkQA commented:
**[Test build #139535 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139535/testReport)** for PR 32303 at commit [`f00f210`](https://github.com/apache/spark/commit/f00f210cafbaae94d0fdf81d44d678172a75b1d1).

cloud-fan committed:
Update ResolveSubquerySuite.scala

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44053/

SparkQA commented:
**[Test build #139540 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139540/testReport)** for PR 32303 at commit [`eb46d13`](https://github.com/apache/spark/commit/eb46d13a2b5696bd0307c4727a69af5a06de92d9).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44059/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44059/

SparkQA commented:
**[Test build #139540 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139540/testReport)** for PR 32303 at commit [`eb46d13`](https://github.com/apache/spark/commit/eb46d13a2b5696bd0307c4727a69af5a06de92d9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139540/

cloud-fan committed:
Update ResolveSubquerySuite.scala

SparkQA commented:
**[Test build #139543 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139543/testReport)** for PR 32303 at commit [`08385f4`](https://github.com/apache/spark/commit/08385f4a0cbba256ae93131100aee16db2feef64).

SparkQA commented:
**[Test build #139543 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139543/testReport)** for PR 32303 at commit [`08385f4`](https://github.com/apache/spark/commit/08385f4a0cbba256ae93131100aee16db2feef64).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44053/

cloud-fan committed:
Update ResolveSubquerySuite.scala

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44053/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139543/

SparkQA commented:
**[Test build #139546 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139546/testReport)** for PR 32303 at commit [`1b6d8f3`](https://github.com/apache/spark/commit/1b6d8f31d2dcce6ebfdd5709b3d61ad5cef9a8b9).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44064/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44064/

SparkQA commented:
**[Test build #139528 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139528/testReport)** for PR 32303 at commit [`8f19ab7`](https://github.com/apache/spark/commit/8f19ab732fa0117a72626c601e964140c9b4bd42).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139528/

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #139562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139562/testReport)** for PR 32303 at commit [`1b6d8f3`](https://github.com/apache/spark/commit/1b6d8f31d2dcce6ebfdd5709b3d61ad5cef9a8b9).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44068/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44068/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44071/

SparkQA commented:
**[Test build #139535 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139535/testReport)** for PR 32303 at commit [`f00f210`](https://github.com/apache/spark/commit/f00f210cafbaae94d0fdf81d44d678172a75b1d1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44071/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139535/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44071/

SparkQA commented:
**[Test build #139546 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139546/testReport)** for PR 32303 at commit [`1b6d8f3`](https://github.com/apache/spark/commit/1b6d8f31d2dcce6ebfdd5709b3d61ad5cef9a8b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139546/

SparkQA commented:
**[Test build #139562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139562/testReport)** for PR 32303 at commit [`1b6d8f3`](https://github.com/apache/spark/commit/1b6d8f31d2dcce6ebfdd5709b3d61ad5cef9a8b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139562/

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44089/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44089/

cloud-fan committed:
Update ResolveSubquerySuite.scala

SparkQA commented:
**[Test build #139577 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139577/testReport)** for PR 32303 at commit [`01fa70e`](https://github.com/apache/spark/commit/01fa70ec156ad5540b862c4b13dd1257062f31fb).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44106/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44106/

cloud-fan commented:
The GA failure is in `Linters, licenses, dependencies and documentation generation`, which is unrelated to this PR. I'm merging it to master, thanks!

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44106/

SparkQA commented:
**[Test build #139577 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139577/testReport)** for PR 32303 at commit [`01fa70e`](https://github.com/apache/spark/commit/01fa70ec156ad5540b862c4b13dd1257062f31fb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139577/
